src/bundle/Security/TwoFactor/Event/ThrottlingListener.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Security\TwoFactor\Event;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Exception\TooManyRequestsHttpException;
+use Symfony\Component\RateLimiter\RateLimiterFactory;
+
+final class ThrottlingListener implements EventSubscriberInterface
+{
+    public function __construct(
+        private RateLimiterFactory $rateLimiterFactory
+    ) {
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            TwoFactorAuthenticationEvents::ATTEMPT => 'onTwoFactorAttempt',
+            TwoFactorAuthenticationEvents::SUCCESS => 'onTwoFactorSuccess',
+        ];
+    }
+
+    public function onTwoFactorAttempt(TwoFactorAuthenticationEvent $event): void
+    {
+        $rateLimiter = $this->rateLimiterFactory->create($event->getRequest()->getClientIp());
+
+        if (!$rateLimiter->consume()->isAccepted()) {
+            throw new TooManyRequestsHttpException();
+        }
+    }
+
+    public function onTwoFactorSuccess(TwoFactorAuthenticationEvent $event): void
+    {
+        $rateLimiter = $this->rateLimiterFactory->create($event->getRequest()->getClientIp());
```

Comment:
This should have some kind of unique keyword in addition to the IP, to avoid conflicts with other pieces of code, e.g.:

```php
private const RATE_LIMITER_KEY_PREFIX = "scheb_two_factor.brute_force_protection.";

$this->rateLimiterFactory->create(self::RATE_LIMITER_KEY_PREFIX  . $event->getRequest()->getClientIp())
```

Reply:
I can do that of course, but I'm not sure if it's necessary, because the rate limiter component will automatically prefix the key with its ID (see https://github.com/symfony/rate-limiter/blob/6.2/RateLimiterFactory.php#L46)

Reply:
Resolved via https://github.com/scheb/2fa/pull/165#discussion_r1038863299, no need to change.